File: swift/BCComponents/Sources/BCComponents/Ed25519PublicKey.swift
import BCCrypto
import Foundation

public struct Ed25519PublicKey: Equatable, Hashable, Sendable {
    public static let keySize = ed25519PublicKeySize

    private let value: Data

    public init(_ value: Data) throws(BCComponentsError) {
        try requireLength(value, expected: Self.keySize, name: "Ed25519 public key")
        self.value = value
    }

    public static func fromHex(_ hex: String) throws(BCComponentsError) -> Ed25519PublicKey {
        try Ed25519PublicKey(parseHex(hex))
    }

    public var data: Data {
        value
    }

    public func verify(
        _ signature: Data,
        _ message: some DataProtocol
    ) -> Bool {
        ed25519Verify(value, Data(message), signature)
    }
}

extension Ed25519PublicKey: ReferenceProvider {
    public func reference() -> Reference {
        Reference.fromDigest(Digest.fromImage(value))
    }
}

extension Ed25519PublicKey: CustomStringConvertible {
    public var description: String {
        "Ed25519PublicKey(\(refHexShort()))"
    }
}

extension Ed25519PublicKey: CustomDebugStringConvertible {
    public var debugDescription: String {
        "Ed25519PublicKey(\(hexEncode(value)))"
    }
}

File: typescript/bc-components/src/ed25519/ed25519-public-key.ts
import {
    ed25519Verify,
    ED25519_PUBLIC_KEY_SIZE,
} from '@bc/crypto';

import { Digest } from '../digest.js';
import { type ReferenceProvider, Reference } from '../reference.js';
import { bytesEqual, hexDecode, hexEncode, requireLength } from '../utils.js';

export class Ed25519PublicKey implements ReferenceProvider {
    static readonly KEY_SIZE = ED25519_PUBLIC_KEY_SIZE;

    readonly #data: Uint8Array;

    constructor(data: Uint8Array) {
        this.#data = requireLength(data, ED25519_PUBLIC_KEY_SIZE, 'Ed25519 public key');
    }

    static fromData(data: Uint8Array): Ed25519PublicKey {
        return new Ed25519PublicKey(new Uint8Array(data));
    }

    static fromDataRef(data: Uint8Array): Ed25519PublicKey {
        return Ed25519PublicKey.fromData(data);
    }

    static fromHex(hex: string): Ed25519PublicKey {
        return Ed25519PublicKey.fromData(hexDecode(hex));
    }

    data(): Uint8Array {
        return new Uint8Array(this.#data);
    }

    asBytes(): Uint8Array {
        return this.data();
    }

    hex(): string {
        return hexEncode(this.#data);
    }

    verify(signature: Uint8Array, message: Uint8Array): boolean {
        return ed25519Verify(this.#data, message, signature);
    }

    reference(): Reference {
        return Reference.fromDigest(Digest.fromImage(this.#data));
    }

    equals(other: unknown): boolean {
        return other instanceof Ed25519PublicKey && bytesEqual(this.#data, other.#data);
    }

    toString(): string {
        return `Ed25519PublicKey(${this.reference().refHexShort()})`;
    }
}
